### ISSUE TITLE
Tab List API

### ISSUE THREAD

**Issue by AlphaModder** [SpongePowered/SpongeAPI#391]
This is my idea for a tab-list API. It's a simple two files and a method in Player.

First file: TabList.java:
```java
package org.spongepowered.api.entity.player.tab;

import java.util.List;
import java.util.UUID;

import org.spongepowered.api.text.message.Message;

import com.google.common.base.Optional;

public interface TabList {
	
	/**
	 * Sets this list's header.
	 * @param header The new header.
	 */
	public void setHeader(Message header);
	
	/**
	 * Gets this list's header.
	 * @return The current header.
	 */
	public Message getHeader();
	
	/**
	 * Sets this list's footer.
	 * @param footer The new footer.
	 */
	public void setFooter(Message footer);
	
	/**
	 * Gets this list's footer.
	 * @return The current footer.
	 */
	public Message getFooter();
	
	/**
	 * Gets the players on the list.
	 * @return The players on the list.
	 */
	public List<PlayerTabInfo> getPlayers();
	
	/**
	 * <p>Adds a player to the list. This should throw an 
	 * IllegalArgumentException when it attempts to add a player
	 * already on the list. This is to prevent modification of a 
	 * {@link PlayerTabInfo} by overwriting it.</p>
	 * @param player The player to add.
	 */
	public void addPlayer(PlayerTabInfo player) throws IllegalArgumentException;
	
	/**
	 * <p>Removes a player from the list. This should only be used to completely
	 * remove a player, not add it back later.</p>
	 * @param playerId the UUID of the player to remove.
	 * @return The {@link PlayerTabInfo} that was associated with the UUID.
	 */
	public PlayerTabInfo removePlayer(UUID playerId);
	
	/**
	 * <p>Finds a {@link PlayerTabInfo} matching the specified UUID. If none
	 * were found, it returns Optional.absent().</p>
	 * @param uuid The UUID to search for.
	 * @return An Optional containing a PlayerTabInfo if one was found.
	 */
	public Optional<PlayerTabInfo> getPlayer(UUID uuid);
	
}

```
Second file: PlayerTabInfo.java
```java
package org.spongepowered.api.entity.player.tab;

import org.spongepowered.api.entity.player.gamemode.GameMode;
import org.spongepowered.api.status.PlayerProfile;
import org.spongepowered.api.text.message.Message;

public interface PlayerTabInfo {
	
	/**
	 * Sets the connection time for this player.
	 * @param milliseconds The new connection time, in milliseconds.
	 */
	public void setConnectionTime(int milliseconds);
	
	/**
	 * Gets the connection time for this player.
	 * @return The connection time for this player.
	 */
	public int getConnectionTime();
	
	/**
	 * Sets this player's gamemode.
	 * @param gamemode The new gamemode.
	 */
	public void setGameMode(GameMode gamemode);
	
	/**
	 * Gets the {@link GameMode} this player is in.
	 * @return The Gamemode this player is in.
	 */
	public GameMode getGameMode();
	
	/**
	 * Sets this player's display name.
	 * @param displayName The new display name.
	 */
	public void setDisplayName(Message displayName);
	
	/**
	 * Gets this player's display name.
	 * @return This player's display name.
	 */
	public Message getDisplayName();
	
	/**
	 * Sets this player's username.
	 * @param name A new username.
	 */
	public void setName(String name);
	
	/**
	 * Gets this player's username.
	 * @return This player's username.
	 */
	public String getName();
	
	/**
	 * Sets the {@link PlayerProfile} associated with this player.
	 * @param profile The new PlayerProfile for this player.
	 */
	public void setProfile(PlayerProfile profile);
	
	/**
	 * Gets the {@link PlayerProfile} associated with this player.
	 * @return The PlayerProfile associated with this player.
	 */
	public PlayerProfile getProfile();
}

```

It would also add a `TabList getTabList()` to Player, for obvious reasons. `TabPlayerInfo` will need some way of being created by plugins, because they may want to add fake tablist entries, and consequently this would require a way to create `PlayerProfile`s. I talked about this a while ago in IRC.

Implementation Details Below:
NOTE: These are MCP names.
Whenever someone calls `setHeader()` / `setFooter()`, the server should send a `S47PacketPlayerListHeaderFooter` containing the new header and footer to that client.

PlayerTabInfo would have to behind the scenes have a is-linked value, that is true when it's in a tablist, and set to false when `removePlayer()` is called, to not only prevent abuse of `removePlayer()` If it's true, when calling setter methods in one a `S38PacketPlayerListItem` must be sent to the client with an appropriate `Action` and data, one of it's subclasses. For reference the Actions that must be sent when modifying a PlayerTabInfo are `UPDATE_GAME_MODE`, `UPDATE_DISPLAY_NAME`, and `UPDATE_LATENCY`. When TabList's `addPlayer()` and `removePlayer()` are called, similarly, the same packet must be sent with an action of `ADD_PLAYER` or `REMOVE_PLAYER`, respectively. The actual info is relatively easy to figure out, just by looking at the packet's code.

**Comment by Aaron1011** [comment]
It should be noted that setting a player's gamemode in the tab list to `SPECTATOR` is responsible for both the player being invisible client-side, **and** noclip being for a player.

**Comment by Aaron1011** [comment]
As the connection time is rendered as a connection bar (the actual value set isn't displayed), the javadoc should indicate how values are mapped to different values.

**Comment by me4502** [comment]
@Aaron1011 Agreed, I just mentioned that over IRC.

**Comment by me4502** [comment]
Also, as you already have done the code.. Why not PR it instead of making an issue?

**Comment by ryantheleach** [comment]
@mcmonkey4eva I noticed you were having loads of issues with players in the tab list on minecraft 1.8.1 maybe you have some thoughts on what should/shouldn't be done.

**Comment by SpaceManiac** [comment]
Be aware that spoofing the tab list to remove players will influence skin rendering if the player is visible in-game, since the tab list is used to transfer skin information.

**Comment by Aaron1011** [comment]
The skin data is actually sent as well, in the form of a property array. However, it does have to be signed by Mojang.

**Comment by AlphaModder** [comment]
@SpaceManaiac Oh, I didn't know that. I'd assume the main use of removing someone would be vanish plugins, however.

**Comment by AlphaModder** [comment]
@me4502 Generally, unless it's more than a few interfaces I open an issue to allow for the simplest feedback, we can discuss edits and just use the original code a template for what we need, and then someone else can conglomerate the suggestions.

**Comment by Aaron1011** [comment]
I'm going to work on this, unless someone else is.

**Comment by gratimax** [comment]
@AlphaModder You seem to have already made most of the changes, this would fit great as a PR

**Comment by AlphaModder** [comment]
@gratimax So, I should make one then?

**Comment by gratimax** [comment]
@AlphaModder yes, that's what I was saying

**Comment by simon816** [comment]
Can be closed now, #433 was merged.
